### PR TITLE
[SourceTree] fails on https://facebook.github.io/react/

### PR DESCRIPTION
**Pull request by codehag**
associated issue: #3430 

screenshot: 
screen shot 2017-08-15 at 8 07 57 pm

### CODE REVIEW

File: src/utils/sources-tree/tests/addToTree.spec.js [hunk 1]
```diff
@@ -295,9 +296,96 @@ describe("sources-tree", () => {
       expect(formatTree(tree)).toMatchSnapshot();
     });
 
-    it("correctly parses file sources correctly", () => {
+    it("doesnt throw when adding a deeper file", () => {
```
Comment (jasonLaster) [inline src/utils/sources-tree/tests/addToTree.spec.js]: maybe we need a better name?

File: src/utils/sources-tree/tests/addToTree.spec.js [hunk 2]
```diff
@@ -295,9 +296,96 @@ describe("sources-tree", () => {
...
+
+      const source2 = Map({
+        url: "http://b/b.js",
+        actor: "actor1"
+      });
+
+      const tree = createNode("root", "", []);
+      addToTree(tree, source1);
+      addToTree(tree, source2);
+      const paths = getDirectories("http://a/b.js?key=hi", tree);
```
Comment (jasonLaster) [inline src/utils/sources-tree/tests/addToTree.spec.js]: this test might be in `getDirectories`...

File: src/utils/sources-tree/tests/source-tree.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      {
+        id: "server1.conn13.child1/39",
+        url: "https://unpkg.com/codemirror/mode/xml/xml.js"
+      }
+    ];
+
+    const testMap = createSourcesList(testData);
+    const tree = createTree(testMap);
+    console.log(tree.uncollapsedTree.contents[0]);
+    console.log("YOOO\n", formatTree(tree.uncollapsedTree));
```
Comment (jasonLaster) [inline src/utils/sources-tree/tests/source-tree.js]: the tests seem a bit funny... like not sure why we need all of them

File: src/utils/sources-tree/addToTree.js [hunk 4]
```diff
@@ -85,48 +87,52 @@ export function addToTree(
...
+      const sourceContents = subtree.contents;
+      const contentsUrl = getURL(sourceContents.get("url"));
+      const name = getFilenameFromPath(contentsUrl.path);
+
+      const nextNode = createNode(part, `${path}/${part}`, []);
+      // this directory has the same name as the file
+      const newDir = createNode(name, `${path}`, [nextNode]);
+      const newSource = createNode(name, `${path}`, source);
+      // append the new directory, with the file and the new directory
+      ancestor.contents = [newDir, newSource];
```
Comment (jasonLaster) [inline src/utils/sources-tree/addToTree.js]: **YES!**

File: src/utils/sources-tree/tests/addToTree.spec.js (hunk 2 above)
Comment (codehag) [inline src/utils/sources-tree/tests/addToTree.spec.js]: :+1: thanks for catching that!

Comment (codehag) [inline src/utils/sources-tree/tests/addToTree.spec.js]: we dont have a getDirectories test at the moment, and it isn't in any of the other source tree test files, so I will leave this here for now

File: src/utils/sources-tree/addToTree.js [hunk 5]
```diff
@@ -85,48 +87,52 @@ export function addToTree(
   url.path = decodeURIComponent(url.path);
 
   const parts = url.path.split("/").filter(p => p !== "");
-  const isDir = isDirectory(url);
   parts.unshift(url.group);
 
   let path = "";
+  let part = "";
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: we don't need this to be declared here

File: src/utils/sources-tree/addToTree.js [hunk 6]
```diff
@@ -85,48 +87,52 @@ export function addToTree(
...
+        subtree = subtree.contents[where];
+      }
     } else {
-      // No node with this name exists, so insert a new one in the
-      // place that is alphabetically sorted.
-      const node = createNode(part, `${path}/${part}`, []);
-      const where = index === -1 ? children.length : index;
-      children.splice(where, 0, node);
-      subtree = children[where];
+      // we enter this else block if a file and a folder have the same
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: at some point we might want to revisit this code, it feels like there is a simpler solution

**Comment by jasonLaster** [comment]
pushed a rebase and a couple small test fixups. should be ready to go when green :)

File: src/utils/sources-tree/tests/__snapshots__/addToTree.spec.js.snap [hunk 7]
```diff
@@ -10,35 +10,27 @@ exports[`sources-tree addToTree alphabetically sorts children 1`] = `
...
 `;
 
-exports[`sources-tree addToTree correctly parses file sources correctly 1`] = `
+exports[`sources-tree addToTree can add a file to an intermediate directory 1`] = `
 " - root path= 
-  - file:// path=/file:// 
-    - a path=/file:///a 
-      - b.js path=/file:///a/b.js source_id=undefined 
+  - unpkg.com path=/unpkg.com 
+    - codemirror path=/unpkg.com/codemirror source_id=server1.conn13.child1/37
```
Comment (jasonLaster) [inline src/utils/sources-tree/tests/__snapshots__/addToTree.spec.js.snap]: this test should also have the xml mode

**Pull request merged**